Pinta.Effects/Effects/OutlineObjectEffect.cs

```diff
@@ -0,0 +1,192 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Cairo;
+using Pinta.Core;
+using Pinta.Gui.Widgets;
+
+namespace Pinta.Effects;
+
+public sealed class OutlineObjectEffect : BaseEffect
+{
+
+	public override string Icon => Pinta.Resources.Icons.EffectsStylizeOutline;
+
+	// Takes two passes, so must be multithreaded internally
+	public sealed override bool IsTileable => false;
+
+	public override string Name => Translations.GetString ("Outline Object");
+
+	public override bool IsConfigurable => true;
+
+	public override string EffectMenuCategory => Translations.GetString ("Object");
+
+	public OutlineObjectData Data => (OutlineObjectData) EffectData!;  // NRT - Set in constructor
+
+	private readonly IChromeService chrome;
+	private readonly ISystemService system;
+	private readonly IPaletteService palette;
+
+	public OutlineObjectEffect (IServiceProvider services)
+	{
+		chrome = services.GetService<IChromeService> ();
+		system = services.GetService<ISystemService> ();
+		palette = services.GetService<IPaletteService> ();
+		EffectData = new OutlineObjectData ();
+	}
+
+	public override void LaunchConfiguration ()
+		=> chrome.LaunchSimpleEffectDialog (this);
+
+	protected override void Render (ImageSurface src, ImageSurface dest, RectangleI roi)
+	{
+		int top = roi.Top;
+		int bottom = roi.Bottom;
+		int left = roi.Left;
+		int right = roi.Right;
+		int srcHeight = src.Height;
+		int srcWidth = src.Width;
+		int radius = Data.Radius;
+		int threads = system.RenderThreads;
+
+		ColorBgra primaryColor = palette.PrimaryColor.ToColorBgra ();
+		ColorBgra secondaryColor = palette.SecondaryColor.ToColorBgra ();
+		ConcurrentBag<PointI> borderPixels = new ConcurrentBag<PointI> ();
+
+		// First pass
+		// Clean up dest, then collect all border pixels
+		Parallel.For (top, bottom + 1, new ParallelOptions { MaxDegreeOfParallelism = threads }, y => {
+			var srcData = src.GetReadOnlyPixelData ();
+
+			// reset dest to src
+			// Removing this causes preview to not update to lower radius levels
+			var srcRow = srcData.Slice (y * srcWidth, srcWidth);
+			var dstRow = dest.GetPixelData ().Slice (y * srcWidth, srcWidth);
+			for (int x = left; x <= right; x++)
```

Comment:
This loop could be replaced by `CopyTo()`

Reply:
Done